File: arrays/Solution_1813752452.java
class Solution {
    public int rob(int[] A) {
        
        HashMap<Integer, Integer> DP = new HashMap<>();
        return max(A, 0, false, DP);
    }

    int max(int A[], int start, boolean selected0, HashMap<Integer, Integer> DP) {
        int N = A.length;
        if(selected0 == true && start == N - 1) {
            // you can't selected last if you select 0th index
        }
        if(start == N) {
            return 0;
        }

        if(DP.containsKey(start)) {
            return DP.get(start);
        }

        int profit = 0;
        for(int i = start; i < N; i++) {
            profit = Math.max(profit, A[i] + max(A, i + 2, i == 0, DP)); // need to skip the adjacent one
        }

        DP.put(start, profit);

        return profit;
    }
}

File: arrays/Solution_1813752452.py
class Solution:

    def __init__(self):
        self.dp = {}

    def maximize(self, A, idx):
        n = len(A)
        if idx >= n:
            return 0

        if idx in self.dp:
            return self.dp[idx]

        mx = 0
        for i in range(idx, n):
            mx = max(mx, A[i] + self.maximize(A, i + 2)) # skip next element

        self.dp[idx] = mx
        return mx

    def rob(self, nums: List[int]) -> int:
        return self.maximize(nums, 0)